xapian-applications/omega/handler_libarchive.cc

```diff
@@ -0,0 +1,122 @@
+/** @file handler_libarchive.cc
+ * @brief Extract text and metadata using libarchive.
+ */
+/* Copyright (C) 2020 Parth Kapadia
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+#include <config.h>
+#include "handler.h"
+
+#include "metaxmlparse.h"
+#include "opendocparse.h"
+#include "str.h"
+
+#include<archive.h>
+#include<archive_entry.h>
+
+#include <cstring>
+
+using namespace std;
+
+bool
+extract(const string& filename,
+	string& dump,
+	string& title,
+	string& keywords,
+	string& author,
+	string& pages,
+	string& error)
+{
+    try {
+	struct archive* archive_obj;
+	struct archive_entry* entry;
+	int status_code;
+	const char* file = filename.c_str();
+	archive_obj = archive_read_new();
+	archive_read_support_filter_all(archive_obj);
+	archive_read_support_format_all(archive_obj);
+	status_code = archive_read_open_filename(archive_obj, file, 10240);
+
+	if (status_code != ARCHIVE_OK) {
+	    error = "Libarchive failed to open the file " + filename;
+		return false;
+	}
+
+	size_t total;
+	ssize_t size;
+	string s;
+
+	// extracting data from content.xml and styles.xml
+	while (archive_read_next_header(archive_obj, &entry) == ARCHIVE_OK) {
+	    if (strcmp(archive_entry_pathname(entry), "content.xml") == 0) {
```

Comment:
Currently the structure of this is:

```
    if (strcmp(archive_entry_pathname(entry), "...") == 0) {
        // ...
    } else if (strcmp(archive_entry_pathname(entry), "...") == 0) {
        // ...
    } else if (strcmp(archive_entry_pathname(entry), "...") == 0) {
        // ...
    }
```

If an entry doesn't match any of these checks (which most probably don't - a typical ODF zip container has a lot more than just these 3 files) or matches the last one then we call `archive_entry_pathname(entry)` 3 times.  It'll give the same answer each time, but the compiler has no way to tell that.

I'd suggest just calling it once and putting the result in a variable.  That'll also make the code much clearer - the conditions will be short enough to fit on one line, and if you use a string variable you can just write them as equality checks:

```
    string pathname = archive_entry_pathname(entry);
    if (pathname == "content.xml") {
        // And so on...
```

Reply:
Thanks for pointing this out. Yes, multiple calls to archive_entry_pathname() were redundant. I have changed this, used `string pathname` instead.